### ISSUE TITLE
`torch.utils.checkpoint` should avoid updating BatchNorm statistics twice

### ISSUE THREAD

**Issue by zhengwy888** [pytorch/pytorch#96136]
### 🚀 The feature, motivation and pitch

`torch.utils.checkpoint` provides two checkpointing methods
1. Reentrant
2. non-reentrant 

In both cases if the wrapped module contains `BatchNorm` layers, the batch norm will be updated twice because the forward pass has to be ran twice. This behavior makes reproducing experiment results between checkpointed and un-checkpointed runs harder. 
To repro the problem, one can run https://github.com/facebookresearch/fairscale/blob/main/tests/nn/checkpoint/test_checkpoint_activations_norm.py test without the fairscale wrapper. 

### Alternatives

Fairscale provides a patched Reentrant flavour of activation checkpoint wrapper, that uses 2 hooks to avoid the second forward pass updating the batchnorm stats. 
But currently we don't have a clean API to support this behavior in the non-reentrant case. I can think of one method to achieve this effect is to make the following modification in the non-reentrant pass, but not sure if this leads to a more complicated code. 
1. mark the `running_mean` and `running_var` tensor inside BatchNorm with a special attribute.
2. detect that special attribute during [pack](https://github.com/pytorch/pytorch/blob/master/torch/utils/checkpoint.py#L378), and return the normal tensor instead of the holder object
3. during `unpack`, if a tensor is passed in as argument, return the tensor directly instead of loading it from `storage`
4. during `unpack`, run the forward pass with `track_running_stats=False`, so the `running_mean` and `running_var` does not get updated.

**Comment by vadimkantorov** [comment]
Not sure, but could this scenario be made more transparent if batchnorm running stats updates were explicit with calling `torch._foreach_lerp_`? I pseudo-coded an example of how it could look like here: https://github.com/pytorch/pytorch/issues/71683#issuecomment-1457094446
